File: Python/lib/upload_s3.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

import s3sync_lib
import os
import codecs
import sys

from boto.s3.connection import S3Connection
from boto.s3.key import Key

config = s3sync_lib.get_config()
input_file = os.path.dirname(__file__) + '/../working/TransferList.txt'

s3conn = S3Connection(config[u'access_key'],config[u'secret_key'])
bucket = s3conn.get_bucket(config[u'bucket'])

with codecs.open(input_file,encoding='UTF-8',mode='r') as file:
    for line in file:
        line = line.strip()
        line_list = line.split('<>')

        action = line_list[0]
        s3key = line_list[1]

        k = Key(bucket)
        k.key = s3key

        if action == 'ADD':
            file_name = line_list[2]
            rrs = line_list[3]
            encrypt = line_list[4]
            mtime = line_list[5]

            if rrs == '1':
                rrs = True
            else:
                rrs = False

            if encrypt == '1':
                encrypt = True
            else:
                encrypt = False

            print('Uploading file %s...' % s3key)

            k.set_contents_from_filename(
                file_name,
                reduced_redundancy=rrs,
                encrypt_key=encrypt
                )

        if action == 'DELETE':
            print('Removing file %s...' % s3key)

            bucket.delete_key(k)

File: PHP/lib/UploadS3.php
<?php

require_once dirname(__FILE__) . '/./common.php';
require_once dirname(__FILE__) . '/../composer/vendor/autoload.php';
use Aws\Common\Enum\Size;
use Aws\Common\Exception\MultipartUploadException;
use Aws\S3\S3Client;
use Aws\S3\Model\MultipartUpload\UploadBuilder;

$handle = fopen(dirname(__FILE__) . '/../working/TransferList.txt','r');
$config = getConfig();

try {
    $client = S3Client::factory(array(
        'key' => $config['access_key'],
        'secret' => $config['secret_key'],
        'region' => $config['region']
    ));

    $doptions = array(
        'Bucket' => $config['bucket']
    );

    while(!feof($handle)) {
        $line = trim(fgets($handle));

        if(strlen($line) > 0) {
            list($path,$key,$op,$rrs,$enc) = explode('<>',$line);

            $uploader = UploadBuilder::newInstance()
                ->setClient($client)
                ->setBucket($config['bucket']);

            if($rrs == 1) {
                $uploader = $uploader->setOption('StorageClass','REDUCED_REDUNDANCY');
            } else {
                $uploader = $uploader->setOption('StorageClass','STANDARD');
            }

            if($enc == 1) {
                $uploader = $uploader->setOption('ServerSideEncryption','AES256');
            }

            if($op === "ADD") {
                echo "Uploading : $path" . PHP_EOL;

                $uploader = $uploader->setSource($path)
                    ->setKey($key)
                    ->build();

                $uploader->upload( );
            } else if($op === "DELETE") {
                echo "Removing : $key" . PHP_EOL;

                $doptions['Key'] = $key;

                $client->deleteObject($doptions);
            }
        }
    }
} catch(Exception $e) {
    echo "Error: " . $e->getMessage() . PHP_EOL;
    exit(1);
}

fclose($handle);
